src_code/src/controllers/services/Reservation/MakeReservation.js

```diff
@@ -0,0 +1,368 @@
+ * @param {Date} endDate - The end date of the reservation.
+ * @returns {string|null} - Error message if validation fails, otherwise null.
+ */
+async function EquipmentValidation(equipment, startDate, endDate) {
+    try {
+        /** Ensure all information of equipment is provided */
+        if (equipment.typeId === undefined || equipment.modelId === undefined || equipment.quantity === undefined || equipment.quantity === null) {
+            return "One of the equipment's information is missing.";
+        } 
+
+        /** Ensure all of the requested typeId, modelId, and quantity is valid type */
+        if(typeof equipment.typeId !== "number" || typeof equipment.modelId !== "number" || typeof equipment.quantity !== "number") {
+            return "Invalid type of an equipment."
+        }
+
+        /** Ensure equipment quantity will be always larger than 0 */
+        if(equipment.quantity === 0) {
+            return "You cannot reserve an equipment with 0 quantity."
+        }
+        
+        /** Ensure that type and model is exist */
+        const typePromise = dbHelpers.GetTypeInfoByTypeId(db, equipment.typeId);
+        const modelPromise = dbHelpers.GetModelInfoByModelId(db, equipment.modelId);
+
+        /** Retrieve available count of the equipment */
+        const availableCountPromise = dbHelpers.GetEquipmentAvailableCount(db, equipment.modelId, equipment.typeId, startDate, endDate);
+        
+        /** Perform the promise concurrently */
+        const [type, model, availableCount] = await Promise.all([typePromise, modelPromise, availableCountPromise])
+
+        /** If type is not exist, return not found. */
+        if(!type) {
+            return "Type not found.";
+        }
+        
+        /** If model is not exist, return not found. */
+        if(!model) {
+            return "Model not found.";
+        }
+
+        /** Extract model ids from type's models */
+        const modelIds = type.models.map(model => model.modelId);
+
+        /** If modelIds doesn't have equipment.modelId, then return "The type doesn't have this model" */
+        if (!modelIds.includes(equipment.modelId)) {
+            return "The type doesn't have this model.";
+        }
+
+        /** Get available number of equipment, if it is error then return error */
+        if(typeof availableCount === "string") {
+            return availableCount;
+        }
+
+        /** If quantity is greater than the sum */
+        if(equipment.quantity > availableCount) {
+            return `The quantity you choose for ${model.modelName} is larger than the available quantity. Please make adjustment.`
+        }
+
+        /** Return null indicate equipment validation is passed */
+        return null;
+    } catch(error) {
+        /** Log and return error */
+        console.log(`ERROR: There is an error while validating equipment with the information \'${equipment}\' for creating reservation:`, error);
```

Comment:
sonarlint: unnecessary escape char \'

![escape](https://github.com/tranleduyan/ETC-Backend/assets/97055517/5fe8eaeb-ab64-44f4-b217-f2ffa4a3b247)

and another sonarlint: `equipment` will evaluate to `[object Object]` when stringified. (Is there a string version of `equipment`)

Reply:
Will change this, it suppose to be equipment.name. thanks

Reply:
Done removing the escape char.